### DISCUSSION TITLE
Storages

### DISCUSSION THREAD

**Discussion by jonnermut** [openfga/community#80] [General]
> OpenFGA's PostgreSQL Storage Adapter was purposely built for OpenFGA. Auth0 is not using it in a production environment.
> OpenFGA's MySQL Storage Adapter was contributed to OpenFGA by [@twintag](https://github.com/twintag). Auth0 is not using it in a production environment.

So what _does_ Auth0 use in its production environment? Another adapter that's not part of the open source?

Has anyone taken the MySQL adapter, or the PostgreSQL for that matter, to its limits? Should we be expecting it to be functioning well with millions or billions of tuples?

**Comment by aaguiarz** [comment]
Hi @jonnermut

Auth0 FGA uses OpenFGA with a Dynamo DB storage adapter, so that multi-region active-active writes can be supported. We are considering open sourcing it too, but we haven’t made a decision yet, initially we wanted the OpenFGA adapters to be for OSS (which DynamoDB is not), which is why we went with Postgres and accepted the MySQL adapter contribution.

The are no theoretical limits on the amount of tuples MySQL/Postgres can handle that are related to how OpenFGA works, and OpenFGA’s core is designed to manage billions of tuples, it is on the implementer to manage scaling, partitioning and sharding the datastore though.

In practice, Auth0 is not using the Postgres or MySQL datastores with a production-like constant workload. Until we get data from production OpenFGA deployments, we won’t know how the product will actually behave under those conditions.

**Comment by tranhl** [comment]
@aaguiarz Has there been any traction on an official DynamoDB storage adapter? My team is really keen on adopting OpenFGA we're a DynamoDB shop so not having an adapter for it is a dealbreaker. Using Postgres is a non-starter as well due to cost/budget constraints, and we don't have the Go expertise to write a custom adapter either.

**Comment by aaguiarz** [comment]
Hi!

We don't have an update yet. What's stopping us is that we need to build a good migration up/down story and that's more challenging than with SQL databases.

**Reply by tranhl** [reply]
Makes sense, there isn't any built-in support for schema migrations DynamoDB. What about just releasing the DB adapter as-is behind an experimental banner? We could ship the adapter without migrations and leave that to the end user as long as the expected schema is documented. We'd be more than happy to take that on as it would enable us to adopt OpenFGA into our stack. Thoughts?

**Comment by aaguiarz** [comment]
I don't think we would do that. We have done a few migrations with DynamoDB ourselves and it goes well beyond defining the new schema.  

We'll be exploring ways to handle schema migrations but we don't have a solid timeline yet.
